### PR TITLE
Network Setup Optimization 

### PR DESCRIPTION
**Pull request by davidvossel**
This optimization prevents attempting to setup phase1 of a vmi multiple times. Phase1 involves reaching into the VMI pods namespace and performing some tasks. The namespace switch involves killing a POSIX thread, which is expensive. We'd rather only do this when it's absolutely necessary.

```release-note
NONE
```

### CODE REVIEW

**Comment by davidvossel** [comment]
@booxter @rmohr this follows up with the network namespace optimization

**Comment by davidvossel** [comment]
@booxter @rmohr this follows up with the network namespace optimization

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -235,6 +243,50 @@ func (d *VirtualMachineController) hasTargetDetectedDomain(vmi *v1.VirtualMachin
 	return false, timeLeft
 }
 
+func (d *VirtualMachineController) clearPodNetworkPhase1(vmi *v1.VirtualMachineInstance) {
+	d.phase1NetworkSetupCacheLock.Lock()
```
Comment (danielBelenky) [inline pkg/virt-handler/vm.go]: Why is this lock required?

File: pkg/virt-handler/vm.go [hunk 2]
```diff
@@ -152,6 +153,13 @@ type VirtualMachineController struct {
 	podIsolationDetector     isolation.PodIsolationDetector
 	containerDiskMounter     *container_disk.Mounter
 	clusterConfig            *virtconfig.ClusterConfig
+
+	// records if pod network phase1 has completed
+	// phase1 involves cycling an entire posix thread
+	// so for performance, knowing phase1 is complete
+	// prevents cycling an unncessary posix thread.
+	phase1NetworkSetupCache     map[types.UID]int
```
Comment (danielBelenky) [inline pkg/virt-handler/vm.go]: maybe use a Boolean or an empty struct instead?

**Comment by danielBelenky** [comment]
Hey, just a question for my understanding and a small suggestion...
/lgtm

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (davidvossel) [reply pkg/virt-handler/vm.go]: the controller's reconcile loop has multiple goroutines active for processing the work queue. The map I used here isn't threadsafe, so if we're going to do concurrent reads/writes, we need a lock to synchronize that.

File: pkg/virt-handler/vm.go (hunk 2 above)
Comment (davidvossel) [reply pkg/virt-handler/vm.go]: i stored the "pid" as the maps value. I cross reference the pid id just to be certain the setup has actually been done. In theory this pid cross examination shouldn't be necessary, i'm just not 100% sure what happens in the event of several failed migrations... there might be a edge case in there where an entry exists for a vmi in the cache from a previous migration attempt.

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (danielBelenky) [reply pkg/virt-handler/vm.go]: thanks

File: pkg/virt-handler/vm.go (hunk 2 above)
Comment (danielBelenky) [reply pkg/virt-handler/vm.go]: thank you for clarifying this

**Comment by danielBelenky** [comment]
@davidvossel thanks for the reply...
Is there any doc I could read about each network setup phase?

File: pkg/virt-handler/vm.go (hunk 2 above)
Comment (vladikr) [inline pkg/virt-handler/vm.go]: There is a thread-safe sync.Map, turns out it's optimized for other use cases.
https://tip.golang.org/pkg/sync/#Map

**Pull request merged**